akamai/data_akamai_cp_code.go

```diff
@@ -30,6 +31,8 @@ func dataSourceCPCode() *schema.Resource {
 }
 
 func dataSourceCPCodeRead(d *schema.ResourceData, meta interface{}) error {
+	setCorrelationID("dataSourceCPCodeRead-" + CreateNonce())
+	PrintLogHeader()
```

Comment:
This logging approach - header and footer log messages wrapping the actual message - unnecessarily increases the log size while also having a problem with concurrency and parallelism making the logs hard to parse and identify lines with the guid. We can not filter on the key since the rows we want are between the start and end tags instead of including them.  
 
A better solution is pushing a GUID tag into individual log statements and/or adopting a log statement pattern convention. I do not want a customer to cut parallelism to one just to achieve readability of the correlations in the log file. 
 
The a couple different ways you can do this :

1. Standardize the format of log lines.  example:
log.Printf(“[%s] [%s] [%s] %s”, logLevel, api, guid, msg)
e.g. log.Printf(“[DEBUG] [GTMv1] [%s] %s”, "DEBUG", "GTMv1", guid, "Opening call to server") with a guid that is a hash or a string like 19872984-be13-321e-a921731dbe9
This would result in a log like like the following 
[DEBUG] [GTMv1] [19872984-be13-321e-a921731dbe9] Opening call to server
2. Create a class to wrap our logging functions to helps provide the formatting above but have method signatures closer to printf in syntax so it is easier to plug in.  It would need to accept the range of formats supported by printf at least and then would build a new logging string.  Problem is this approach makes it harder to append guid later in the string and is more likely to look like "[guid] [level] [api] msg" since it is far safer to add a new tag at front than it is to edit strings of unknown format.
3. Create a class to standardize logging but instead of making it mirror printf you just rely on replacing existing printf statements to migrate to the new approach.  This is probably the safest approach as it allows us to get the formatting of choice 1 above and groups all the logging code together to make reuse possible.  But eliminates the overhead of making it a plug in replacement for printf relying instead on upfront work to migrate to the new approach.

Reply:
I disagree with the comment that the header/footer approach unnecessarily increases log size. Log size was bound to increase incrementally with the introduction of correlations. But I agree that header/footer does not fit the parallelism inherent to the provider and TF. 

Once I generated logs with this branch, I found that reading off the header/footer information versus the log statements themselves to be counter-intuitive. As a developer, I am used to seeing the key fields of a log statement in the statement itself, going from left to right. So agree that introducing correlation ID into the log statement is a better convention.

I also agree that we need to define a log statement convention as you describe with #1. I would still place [level][api][guid]<message> in that order. The #2 option is not incompatible with #1, in that sense that you describe a Java LogAppender style implementation, to which the log format can be fed in as  regex.

Question is whether Hashicorp / Terraform providers follow a particular convention for log statements or its entirely up to individual providers to choose.

Reply:
The one drawback I see of the log statement embedded GUID approach is that there will no longer be an explicit demarcation of start/finish or header/footer, but a grouping of a set of log statements. Editors' search functions would have to be used to find the first and last occurrence. I expect users will get used to it

Reply:
I would prefer prefix as it makes it easy to cat logifile.txt|grep GUID|remove GUID prefix and you have log file raw and sequential also if we used common logging mechanism that is used in TF provider Debug content would be in same TF log file versus in 2 places (I used to do this in feature branch to test AI calls for bugs)